File: src/main/scala/com/dyonovan/neotech/common/tiles/machines/generators/TileSolarPanel.scala
package com.dyonovan.neotech.common.tiles.machines.generators

import java.text.NumberFormat
import java.util.Locale

import cofh.api.energy.IEnergyReceiver
import com.dyonovan.neotech.collections.EnumInputOutputMode
import com.dyonovan.neotech.common.tiles.MachineGenerator
import com.teambr.bookshelf.api.waila.Waila
import net.minecraft.client.Minecraft
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.EnumFacing

/**
  * This file was created for NeoTech
  *
  * NeoTech is licensed under the
  * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
  * http://creativecommons.org/licenses/by-nc-sa/4.0/
  *
  * @author Dyonovan
  * @since 2/3/2016
  */
class TileSolarPanel extends MachineGenerator with Waila {

    // Hold the current tier
    var tier = 0

    // Secondary Constructor to allow tier
    def this(t: Int) {
        this()
        tier = t
        initEnergy(t)
    }

    /**
      * Set the energy to what mode it should be based on tier
      * @param t The tier number, 1, 2, 3
      */
    def initEnergy(t: Int): Unit = {
        t match {
            case 1 =>
                setMaxEnergyStored(amountEnergy(t))
                setMaxExtract(2000)
                setMaxReceive(2000)
            case 2 =>
                setMaxEnergyStored(amountEnergy(t))
                setMaxExtract(10000)
                setMaxReceive(10000)
            case 3 =>
                setMaxEnergyStored(amountEnergy(t))
                setMaxExtract(100000)
                setMaxReceive(100000)
            case _ =>
        }
        if (worldObj != null)
            worldObj.markBlockForUpdate(pos)
    }

    /**
      * How much energy is expected by tier
      * @param t The tier
      * @return How much should be stored
      */
    def amountEnergy(t: Int): Int = {
        t match {
            case 1 => 25000
            case 2 => 1000000
            case 3 => 10000000
            case _ => 0
        }
    }

    /**
      * The initial size of the inventory
      *
      * @return
      */
    override def initialSize: Int = 0

    /**
      * Add all modes you want, in order, here
      */
    def addValidModes() : Unit = {}

    /**
      * We don't need IO here
      * @return False to prevent
      */
    override def shouldHandleIO = false

    /**
      * This method handles how much energy to produce per tick
      *
      * @return How much energy to produce per tick
      */
    override def getEnergyProduced: Int = {
        tier match {
            case 1 => 8
            case 2 => 32
            case 3 => 128
            case _ => 0
        }
    }

    /**
      * Called to tick generation. This is where you add power to the generator
      */
    override def generate(): Unit = energyStorage.receiveEnergy(generating(), false)

    /**
      * Called per tick to manage burn time. You can do nothing here if there is nothing to generate. You should decrease burn time here
      * You should be handling checks if burnTime is 0 in this method, otherwise the tile won't know what to do
      *
      * @return True if able to continue generating
      */
    override def manageBurnTime(): Boolean = {
        if (worldObj.canSeeSky(pos) && worldObj.getSunBrightnessFactor(1.0F) > 0.7F) return true

        false
    }

    /*******************************************************************************************************************
      **************************************************  Tile Methods  ************************************************
      ******************************************************************************************************************/

    override def getSupposedEnergy : Int = { amountEnergy(tier) }

    override def doWork(): Unit = {
        didWork = burnTime == 1

        //Transfer
        if (energyStorage.getEnergyStored > 0) {
                worldObj.getTileEntity(pos.offset(EnumFacing.DOWN)) match {
                    case tile: IEnergyReceiver =>
                        val want = tile.receiveEnergy(EnumFacing.DOWN, energyStorage.getEnergyStored, true)
                        if (want > 0) {
                            val actual = extractEnergy(EnumFacing.UP, want, simulate = false)
                            tile.receiveEnergy(EnumFacing.DOWN, actual, false)
                            didWork = true
                        }
                    case _ =>
                }
        }

        if(manageBurnTime()) {
            generate()
            didWork = true
        } else
            reset()

        if (didWork) {
            worldObj.markBlockForUpdate(pos)
        }
    }


    private def generating(): Int = {
        if (worldObj.canSeeSky(pos) && worldObj.getSunBrightnessFactor(1.0F) > 0.7F) {
            val light = worldObj.getSunBrightnessFactor(1.0F)
            (light * getEnergyProduced).toInt
        } else 0
    }

    /**
      * This will try to take things from other inventories and put it into ours
      */
    override def tryInput() : Unit =  { /* No Op, no input */ }

    /**
      * This will try to take things from our inventory and try to place them in others
      */
    override def tryOutput(): Unit = { /* No Op, no output */ }

    /**
      * Write the tag
      */
    override def writeToNBT(tag: NBTTagCompound): Unit = {
        super[MachineGenerator].writeToNBT(tag)
        tag.setInteger("Tier", tier)
    }

    /**
      * Read the tag
      */
    override def readFromNBT(tag: NBTTagCompound): Unit = {
        super[MachineGenerator].readFromNBT(tag)
        if (tag.hasKey("Tier")) {
            if (tier == 0) initEnergy(tag.getInteger("Tier"))
            tier = tag.getInteger("Tier")
        }
        tier = tag.getInteger("Tier")
    }

    /*******************************************************************************************************************
      ************************************************ Inventory methods ***********************************************
      ******************************************************************************************************************/

    /**
      * Used to get what slots are allowed to be input
      *
      * @return The slots to input from
      */
    override def getInputSlots(mode : EnumInputOutputMode) : Array[Int] = Array()

    /**
      * Used to get what slots are allowed to be output
      *
      * @return The slots to output from
      */
    override def getOutputSlots(mode : EnumInputOutputMode) : Array[Int] = Array()

    override def canExtractItem(index: Int, stack: ItemStack, direction: EnumFacing): Boolean = false
    override def canInsertItem(slot: Int, itemStackIn: ItemStack, direction: EnumFacing): Boolean = false

    /*******************************************************************************************************************
      ************************************************** Energy methods ************************************************
      ******************************************************************************************************************/

    override def canConnectEnergy(from: EnumFacing): Boolean = from == EnumFacing.DOWN

    /*******************************************************************************************************************
      *************************************************** Misc methods *************************************************
      ******************************************************************************************************************/

    /**
      * Used to output the redstone single from this structure
      *
      * Use a range from 0 - 16.
      *
      * 0 Usually means that there is nothing in the tile, so take that for lowest level. Like the generator has no energy while
      * 16 is usually the flip side of that. Output 16 when it is totally full and not less
      *
      * @return int range 0 - 16
      */
    override def getRedstoneOutput: Int = (energyStorage.getEnergyStored * 16) / energyStorage.getMaxEnergyStored

    /**
      * Used to get what particles to spawn. This will be called when the tile is active
      */
    override def spawnActiveParticles(xPos: Double, yPos: Double, zPos: Double): Unit = { }

    /**
      * Waila
      */
    override def returnWailaBody(tipList: java.util.List[String]) : java.util.List[String] = {
        tipList.add("Generating: " + generating() + " (" + (if (generating() == 0) 0 else (worldObj.getSunBrightnessFactor(1.0F) * 100).toInt) + "%)")
        tipList.add("Max: " + getEnergyProduced)
        val stored = NumberFormat.getNumberInstance(Locale.forLanguageTag(Minecraft.getMinecraft.gameSettings.language))
          .format(energyStorage.getEnergyStored)
        val max = NumberFormat.getNumberInstance(Locale.forLanguageTag(Minecraft.getMinecraft.gameSettings.language))
          .format(energyStorage.getMaxEnergyStored)
        tipList.add(stored + "/" + max)
        tipList
    }
}

File: src/main/java/com/teambrmodding/neotech/common/tiles/machines/generators/TileSolarPanel.java
package com.teambrmodding.neotech.common.tiles.machines.generators;

import com.teambr.bookshelf.util.EnergyUtils;
import com.teambrmodding.neotech.collections.EnumInputOutputMode;
import com.teambrmodding.neotech.common.tiles.MachineGenerator;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.EnumFacing;
import net.minecraft.world.World;
import net.minecraftforge.energy.CapabilityEnergy;

/**
 * This file was created for NeoTech
 *
 * NeoTech is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Paul Davis - pauljoda
 * @since 2/16/2017
 */
public class TileSolarPanel extends MachineGenerator {
    public static int BASE_MAX_STORED      = 32000;
    public static int BASE_ENERGY_CHANGE   = 2000;
    public static int BASE_ENERGY_PRODUCED = 16;
    protected int tier;

    public TileSolarPanel() {

    }

    public TileSolarPanel(int tier) {
        this.tier = tier;
        initEnergy();
    }

    /**
     * Sets the default energy settings based on tier
     */
    protected void initEnergy() {
        switch (tier) {
            case 3:
                energyStorage.setMaxExtract(BASE_ENERGY_CHANGE * 8);
                energyStorage.setMaxInsert(BASE_ENERGY_CHANGE * 8);
            case 2:
                energyStorage.setMaxExtract(BASE_ENERGY_CHANGE * 4);
                energyStorage.setMaxInsert(BASE_ENERGY_CHANGE * 4);
                break;
            case 1:
            default :
                energyStorage.setMaxExtract(BASE_ENERGY_CHANGE);
                energyStorage.setMaxInsert(BASE_ENERGY_CHANGE);
        }
        energyStorage.setMaxStored(maxStored());
    }

    /**
     * Gets what the max storage for this tier should be
     * @return
     */
    protected int maxStored() {
        switch (tier) {
            case 3 :
                return BASE_MAX_STORED * 8;
            case 2 :
                return BASE_MAX_STORED * 4;
            case 1:
            default:
                return BASE_MAX_STORED;
        }
    }

    /**
     * The initial size of the inventory
     */
    @Override
    public int getInitialSize() {
        return 0;
    }

    /**
     * Add all modes you want, in order, here
     */
    @Override
    public void addValidModes() {

    }

    /**
     * Used to specify if this tile should handle IO, and render in GUI
     *
     * @return False to prevent
     */
    @Override
    public boolean shouldHandleIO() {
        return false;
    }

    /**
     * This method handles how much energy to produce per tick
     *
     * @return How much energy to produce per tick
     */
    @Override
    public int getEnergyProduced() {
        float lightModifier = 0;
        if(worldObj.canSeeSky(pos) && worldObj.getSunBrightnessFactor(1.0F) > 0.7F)
            lightModifier = worldObj.getSunBrightnessFactor(1.0F);
        switch (tier) {
            case 3:
                return (int) ((BASE_ENERGY_PRODUCED * 8) * lightModifier);
            case 2:
                return (int) ((BASE_ENERGY_PRODUCED * 4) * lightModifier);
            case 1:
            default:
                return (int) (BASE_ENERGY_PRODUCED * lightModifier);
        }
    }

    /**
     * Called to tick generation. This is where you add power to the generator
     */
    @Override
    public void generate() {
        energyStorage.receivePower(getEnergyProduced(), true);
    }

    /**
     * Called per tick to manage burn time. You can do nothing here if there is nothing to generate. You should decrease burn time here
     * You should be handling checks if burnTime is 0 in this method, otherwise the tile won't know what to do
     *
     * @return True if able to continue generating
     */
    @Override
    public boolean manageBurnTime() {
        return worldObj.canSeeSky(pos) && worldObj.getSunBrightnessFactor(1.0F) > 0.7F;
    }

    /*******************************************************************************************************************
     * Tile Methods                                                                                                    *
     *******************************************************************************************************************/

    /**
     * Used to determine how much energy should be in this tile
     *
     * @return How much energy should be available
     */
    @Override
    public int getSupposedEnergy() {
        return maxStored();
    }


    /**
     * This will try to take things from other inventories and put it into ours
     */
    @Override
    public void tryInput() {
        // No Op
    }

    /**
     * This will try to take things from our inventory and try to place them in others
     */
    @Override
    public void tryOutput() {
        if(energyStorage.getEnergyStored() > 0) {
            EnergyUtils.distributePowerToFaces(this.getCapability(CapabilityEnergy.ENERGY, null),
                    worldObj, pos, energyStorage.getMaxExtract(), false);
        }
    }

    @Override
    public NBTTagCompound writeToNBT(NBTTagCompound compound) {
        super.writeToNBT(compound);
        compound.setInteger("Tier", tier);
        return compound;
    }

    @Override
    public void readFromNBT(NBTTagCompound compound) {
        super.readFromNBT(compound);
        if(compound.hasKey("Tier")) {
            if(tier == 0) {
                tier = compound.getInteger("Tier");
                initEnergy();
            }
        }
        tier = compound.getInteger("Tier");
    }

    /*******************************************************************************************************************
     * Inventory Methods                                                                                               *
     *******************************************************************************************************************/

    /**
     * Used to get what slots are allowed to be input
     *
     * @param mode
     * @return The slots to input from
     */
    @Override
    public int[] getInputSlots(EnumInputOutputMode mode) {
        return new int[0];
    }

    /**
     * Used to get what slots are allowed to be output
     *
     * @param mode
     * @return The slots to output from
     */
    @Override
    public int[] getOutputSlots(EnumInputOutputMode mode) {
        return new int[0];
    }

    /**
     * Used to define if an item is valid for a slot
     *
     * @param index The slot id
     * @param stack The stack to check
     * @return True if you can put this there
     */
    @Override
    public boolean isItemValidForSlot(int index, ItemStack stack) {
        return false;
    }

    /**
     * Can insert the item into the inventory
     *
     * @param slot        The slot
     * @param itemStackIn The stack to insert
     * @param dir         The dir
     * @return True if can insert
     */
    @Override
    public boolean canInsertItem(int slot, ItemStack itemStackIn, EnumFacing dir) {
        return false;
    }

    /**
     * Can this extract the item
     *
     * @param slot  The slot
     * @param stack The stack
     * @param dir   The dir
     * @return True if can extract
     */
    @Override
    public boolean canExtractItem(int slot, ItemStack stack, EnumFacing dir) {
        return false;
    }

    /*******************************************************************************************************************
     * Misc Methods                                                                                                    *
     *******************************************************************************************************************/

    /**
     * Return the container for this tile
     *
     * @param id       Id, probably not needed but could be used for multiple guis
     * @param player   The player that is opening the gui
     * @param worldObj The worldObj
     * @param x        X Pos
     * @param y        Y Pos
     * @param z        Z Pos
     * @return The container to open
     */
    @Override
    public Object getServerGuiElement(int id, EntityPlayer player, World worldObj, int x, int y, int z) {
        return null;
    }

    /**
     * Return the gui for this tile
     *
     * @param id       Id, probably not needed but could be used for multiple guis
     * @param player   The player that is opening the gui
     * @param worldObj The worldObj
     * @param x        X Pos
     * @param y        Y Pos
     * @param z        Z Pos
     * @return The gui to open
     */
    @Override
    public Object getClientGuiElement(int id, EntityPlayer player, World worldObj, int x, int y, int z) {
        return null;
    }

    /**
     * Used to get the description to display on the tab
     *
     * @return The long string with the description
     */
    @Override
    public String getDescription() {
        return null;
    }

    /**
     * Used to output the redstone single from this structure
     *
     * Use a range from 0 - 16.
     *
     * 0 Usually means that there is nothing in the tile, so take that for lowest level. Like the generator has no energy while
     * 16 is usually the flip side of that. Output 16 when it is totally full and not less
     *
     * @return int range 0 - 16
     */
    @Override
    public int getRedstoneOutput() {
        return (energyStorage.getEnergyStored() * 16) / energyStorage.getMaxEnergyStored();
    }

    /**
     * Used to get what particles to spawn. This will be called when the tile is active
     *
     * @param xPos
     * @param yPos
     * @param zPos
     */
    @Override
    public void spawnActiveParticles(double xPos, double yPos, double zPos) {

    }
}
